packages/yew-macro/src/function_component.rs

```diff
@@ -233,8 +233,158 @@ impl FunctionComponent {
             where_clause.predicates.push(bound);
         }
 
+        where_clause.predicates.push(parse_quote! { Self: 'static });
+
         generics
     }
+
+    /// Prints the impl fn.
+    fn print_inner_fn(&self) -> TokenStream {
+        let name = self.inner_fn_ident();
+        let FunctionComponent {
+            ref fn_token,
+            ref attrs,
+            ref block,
+            ref return_type,
+            ref generics,
+            ref arg,
+            ..
+        } = self;
+        let mut block = *block.clone();
+        let (impl_generics, _ty_generics, where_clause) = generics.split_for_impl();
+
+        // We use _ctx here so if the component does not use any hooks, the usused_vars lint will not
+        // be triggered.
+        let ctx_ident = Ident::new("_ctx", Span::mixed_site());
+
+        let mut body_rewriter = BodyRewriter::new(ctx_ident.clone());
+        visit_mut::visit_block_mut(&mut body_rewriter, &mut block);
+
+        quote! {
+            #(#attrs)*
+            #fn_token #name #impl_generics (#ctx_ident: &mut ::yew::functional::HookContext, #arg) -> #return_type
+            #where_clause
+            {
+                #block
+            }
+        }
+    }
+
+    fn print_base_component_impl(&self) -> TokenStream {
+        let component_name = self.component_name();
+        let props_type = &self.props_type;
+        let static_comp_generics = self.create_static_component_generics();
+
+        let (impl_generics, ty_generics, where_clause) = static_comp_generics.split_for_impl();
+
+        // TODO: replace with blanket implementation when specialisation becomes stable.
+        quote! {
+            #[automatically_derived]
+            impl #impl_generics ::yew::html::BaseComponent for #component_name #ty_generics #where_clause {
+                type Message = ();
+                type Properties = #props_type;
+
+                #[inline]
+                fn create(ctx: &::yew::html::Context<Self>) -> Self {
+                    Self {
+                        _marker: ::std::marker::PhantomData,
+                        function_component: ::yew::functional::FunctionComponent::<Self>::new(ctx),
+                    }
+                }
+
+                #[inline]
+                fn update(&mut self, _ctx: &::yew::html::Context<Self>, _msg: Self::Message) -> ::std::primitive::bool {
+                    true
```

Comment:
I actually tried this approach locally and I feel like unless it implements `BaseComponent` (which you cannot do so due to `Context<Self>`) you need comments to justify why there're methods that is technically "const" or with unused variables.

I prefer code without need of explanation than with explanation.

Reply:
It needed implementation for, because when I first saw the impl for `rendered` and `destroy`, I wondered why `self` wasn't mut. I also wasn't sure if you changed the default impl of `changed`, since it's now in a different file. Reading only parts of the implementation, there's no indication where the rest of it is. The less reasoning is in the macro part, the happier I am :)

See https://github.com/WorldSEnder/yew/commit/d2ba3483cea5e57847b9dcb579d0270084969092 for a working sample of what I'm proposing.

Reply:
My two cents: I don't see how it's better to forward one-line functions that just return a literal. Should any more functionality be needed, these functions can then be updated. We also have the option of having a shim which implements the function but doesn't have the same signature. We can convert the data as needed before passing it to the function which has the actual implementation.

Reply:
I'm just saying this since these methods are part of the [public API](https://yew-rs-api--pr2579-direct-base-comp-xdhqfmc0.web.app/next/yew/functional/struct.FunctionComponent.html) currently. I'm fine with "practical" shims if they're changed to `#[doc(hidden)]` or otherwise marked as not part of the public interface and subject to change.

Reply:
I think a note a documentation would do the job. It's pretty similar to the Routable trait in the router